File: src/algorithms/linked-lists/traversal/find-node-by-value/sources/find-node-by-value_test.py
import importlib

module = importlib.import_module("find-node-by-value")
find_node_by_value = module.find_node_by_value
ListNode = module.ListNode


def build_list(values):
    head = None
    for val in reversed(values):
        head = ListNode(val, head)
    return head


def test_found_at_head():
    result = find_node_by_value(build_list([5, 2, 3, 4]), 5)
    assert result is not None
    assert result.value == 5


def test_found_in_middle():
    result = find_node_by_value(build_list([1, 2, 7, 4, 5]), 7)
    assert result is not None
    assert result.value == 7


def test_found_at_end():
    result = find_node_by_value(build_list([1, 2, 3, 9]), 9)
    assert result is not None
    assert result.value == 9


def test_not_found():
    result = find_node_by_value(build_list([1, 2, 3, 4]), 42)
    assert result is None


def test_empty_list():
    result = find_node_by_value(None, 5)
    assert result is None


def test_single_node_match():
    result = find_node_by_value(build_list([42]), 42)
    assert result is not None
    assert result.value == 42


def test_single_node_no_match():
    result = find_node_by_value(build_list([42]), 7)
    assert result is None


if __name__ == "__main__":
    test_found_at_head()
    test_found_in_middle()
    test_found_at_end()
    test_not_found()
    test_empty_list()
    test_single_node_match()
    test_single_node_no_match()
    print("All tests passed.")

File: src/algorithms/linked-lists/traversal/find-node-by-value/__tests__/find-node-by-value_test.rs
include!("../sources/find-node-by-value.rs");

fn build_list(values: &[i32]) -> Option<Box<ListNode>> {
    let mut head: Option<Box<ListNode>> = None;
    for &val in values.iter().rev() {
        head = Some(Box::new(ListNode { value: val, next: head }));
    }
    head
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_found_at_head() {
        let list = build_list(&[5, 2, 3, 4]);
        let result = find_node_by_value(list.as_deref(), 5);
        assert!(result.is_some());
        assert_eq!(result.unwrap().value, 5);
    }

    #[test]
    fn test_found_in_middle() {
        let list = build_list(&[1, 2, 7, 4, 5]);
        let result = find_node_by_value(list.as_deref(), 7);
        assert!(result.is_some());
        assert_eq!(result.unwrap().value, 7);
    }

    #[test]
    fn test_found_at_end() {
        let list = build_list(&[1, 2, 3, 9]);
        let result = find_node_by_value(list.as_deref(), 9);
        assert!(result.is_some());
        assert_eq!(result.unwrap().value, 9);
    }

    #[test]
    fn test_not_found() {
        let list = build_list(&[1, 2, 3, 4]);
        let result = find_node_by_value(list.as_deref(), 42);
        assert!(result.is_none());
    }

    #[test]
    fn test_empty_list() {
        let result = find_node_by_value(None, 5);
        assert!(result.is_none());
    }

    #[test]
    fn test_single_node_match() {
        let list = build_list(&[42]);
        let result = find_node_by_value(list.as_deref(), 42);
        assert!(result.is_some());
        assert_eq!(result.unwrap().value, 42);
    }

    #[test]
    fn test_single_node_no_match() {
        let list = build_list(&[42]);
        let result = find_node_by_value(list.as_deref(), 7);
        assert!(result.is_none());
    }
}
